src/esp32/encoder.rs

```diff
@@ -0,0 +1,282 @@
+use super::pin::PinExt;
+
+use core::ffi::{c_short, c_ulong};
+use esp_idf_hal::gpio::{AnyInputPin, Input, PinDriver};
+use esp_idf_sys as espsys;
+use espsys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_DECREASE as pcnt_count_dec;
+use espsys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_INCREASE as pcnt_count_inc;
+use espsys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_INVERSE as pcnt_mode_reverse;
+use espsys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_KEEP as pcnt_mode_keep;
+use espsys::pcnt_channel_t_PCNT_CHANNEL_0 as pcnt_channel_0;
+use espsys::pcnt_channel_t_PCNT_CHANNEL_1 as pcnt_channel_1;
+use espsys::pcnt_config_t;
+use espsys::pcnt_evt_type_t_PCNT_EVT_H_LIM as pcnt_evt_h_lim;
+use espsys::pcnt_evt_type_t_PCNT_EVT_L_LIM as pcnt_evt_l_lim;
+use espsys::{esp, EspError, ESP_ERR_INVALID_STATE, ESP_OK};
+
+use std::collections::BTreeMap;
+use std::sync::atomic::{AtomicI32, Ordering};
+use std::sync::{Arc, Mutex};
+
+use crate::common::board::BoardType;
+use crate::common::config::{Component, ConfigType};
+use crate::common::encoder::{
+    Encoder, EncoderPosition, EncoderPositionType, EncoderSupportedRepresentations, EncoderType,
+};
+use crate::common::registry::ComponentRegistry;
+use crate::common::status::Status;
+
+use embedded_hal::digital::v2::InputPin;
+
+pub(crate) fn register_models(registry: &mut ComponentRegistry) {
+    if registry
+        .register_encoder(
+            "incremental",
+            &Esp32Encoder::<
+                PinDriver<'_, AnyInputPin, Input>,
+                PinDriver<'_, AnyInputPin, Input>,
+            >::from_config,
+        )
+        .is_err()
+    {
+        log::error!("incremental model is already registered")
+    }
+}
+
+pub struct PulseStorage {
+    pub acc: Arc<AtomicI32>,
+    pub unit: u32,
+}
+
+pub struct Esp32Encoder<A, B> {
+    pulse_counter: Box<PulseStorage>,
+    config: pcnt_config_t,
+    a: A,
+    b: B,
+}
+
+impl<A, B> Esp32Encoder<A, B>
+where
+    A: InputPin + PinExt,
+    B: InputPin + PinExt,
+{
+    pub fn new(a: A, b: B) -> anyhow::Result<Self> {
+        let pcnt = Box::new(PulseStorage {
+            acc: Arc::new(AtomicI32::new(0)),
+            unit: 0,
+        });
+        let mut enc = Esp32Encoder {
+            pulse_counter: pcnt,
+            config: pcnt_config_t {
+                pulse_gpio_num: a.pin(),
+                ctrl_gpio_num: b.pin(),
+                pos_mode: pcnt_count_inc,
```

Comment:
let's create a ticket to make it so the config is passed as an argument should make a new struct with sane default for that too